File: java/common/MathUtil.java
/* Copyright (C) 2015 Zach Ohara
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package common;

import java.util.ArrayList;
import java.util.List;

public class MathUtil {
	
	public static void main(String[] args) {
//		System.out.println(isPrime(4));
//		System.out.println("factors: " + factorsOf(100));
		System.out.println(primeFactorsOf(13195));
	}
	
	public static List<Integer> fibonacciTo(int cap) {
		List<Integer> fibList = new ArrayList<Integer>();
		fibList.add(1);
		fibList.add(2);
		while (fibList.get(fibList.size() - 1) < cap) {
			fibList.add(fibList.get(fibList.size() - 1) + fibList.get(fibList.size() - 2));
		}
		fibList.remove(fibList.size() - 1);
		return fibList;
	}
	
	 public static List<Integer> factorsOf(int n) {
		 List<Integer> factorList = new ArrayList<Integer>();
		 int div = 1;
		 while (!factorList.contains(div)) {
			 if (n % div == 0) {
				 factorList.add(div);
				 if (div * div != n) {
					 factorList.add(n / div);
				 }
			 }
			 div ++;
		 }
		 return factorList;
	 }
	 
	 public static boolean isPrime(int n) {
		 return factorsOf(n).size() == 2;
	 }
	 
	 public static List<Integer> primeFactorsOf(int n) {
		 List<Integer> primeFactorList = factorsOf(n);
		 primeFactorList.remove((Object) n);
		 primeFactorList.remove((Object) 1);
		 for (int i = primeFactorList.size() - 1; i >= 0; i--) {
			 if (!isPrime(primeFactorList.get(i))) {
				 primeFactorList.remove(i);
			 }
		 }
		 return primeFactorList;
	 }
		
//		public static List<BigInteger> fibonacciTo(BigInteger cap) {
//			List<BigInteger> fibList = new ArrayList<BigInteger>();
//			fibList.add(BigInteger.valueOf(1));
//			fibList.add(BigInteger.valueOf(2));
//			while (fibList.get(fibList.size() - 1).compareTo(cap) < 0) {
//				fibList.add(fibList.get(fibList.size() - 1).add(fibList.get(fibList.size() - 2)));
//			}
//			fibList.remove(fibList.size() - 1);
//			return fibList;
//		}
	
//	 public static List<BigInteger> factorsOf(BigInteger n) {
//		 List<BigInteger> factorList = new ArrayList<BigInteger>();
//		 BigInteger div = BigInteger.ONE;
//		 while (!factorList.contains(div)) {
//			 if (n.mod(div).equals(BigInteger.ZERO)) {
//				 factorList.add(div);
//				 if (!div.pow(2).equals(n)) {
//					 factorList.add(n.divide(div));
//				 }
//			 }
//			 div.add(BigInteger.ONE);
//		 }
//		 return factorList;
//	 }
	 
//	 public static boolean isPrime(BigInteger n) {
//		 return factorsOf(n).size() == 2;
//	 }
	 
//	 public static List<BigInteger> primeFactorsOf(BigInteger n) {
//		 List<BigInteger> primeFactorList = factorsOf(n);
//		 primeFactorList.remove((Object) n);
//		 primeFactorList.remove((Object) 1);
//		 for (int i = primeFactorList.size() - 1; i >= 0; i--) {
//			 if (!isPrime(primeFactorList.get(i))) {
//				 primeFactorList.remove(i);
//			 }
//		 }
//		 return primeFactorList;
//	 }
	
}

File: python/common/MathUtil.py

def fibonacciTo(cap):
	fibList = [1, 2]
	while fibList[-1] < cap:
		fibList.append(fibList[-1] + fibList[-2])
	fibList = fibList[:-1]
	return fibList

def factorsOf(n):
	if n == 1:
		return [1]
	factorList = []
	div = 1
	while not div in factorList:
		if n % div == 0:
			factorList.append(div)
			if div ** 2 != n:
				factorList.append(int(n / div))
		div += 1
	return factorList

def isPrime(n):
	return len(factorsOf(n)) == 2

def primeFactorsOf(n):
	primeFactorList = factorsOf(n)
	primeFactorList.remove(1)
	primeFactorList.remove(n)
	for i in primeFactorList[::-1]:
		if not isPrime(i):
			primeFactorList.remove(i)
	return primeFactorList

def generatePrimes(n):
	primes = []
	i = 2
	while len(primes) < n:
		if not containsFactor(primes[1:], i):
			primes.append(i)
		i += 1
	return primes

def primesTo_old(n):
	primes = [2]
	primes.extend(list(range(3, n, 2)))
	print(primes)
	p = primes[0]
	confirmed = 1
	while p ** 2 < n:
		i = p * 2
		while i < n:
			if i in primes:
				primes.remove(i)
				print("removing", i)
			i += p 
		p = primes[confirmed]
		confirmed += 1
		print(p**2)
	return primes

# source: http://stackoverflow.com/a/27215801/5131653
def primesTo(limit):
	P = [2,3]
	sieve=[False]*(limit+1)
	for x in range(1,int(limit ** 0.5)+1):
		for y in range(1,int(limit ** 0.5)+1):
			n = 4*x**2 + y**2
			if n<=limit and (n%12==1 or n%12==5) : sieve[n] = not sieve[n]
			n = 3*x**2+y**2
			if n<= limit and n%12==7 : sieve[n] = not sieve[n]
			n = 3*x**2 - y**2
			if x>y and n<=limit and n%12==11 : sieve[n] = not sieve[n]
	for x in range(5,int(limit ** 0.5)):
		if sieve[x]:
			for y in range(x**2,limit+1,x**2):
				sieve[y] = False
	for p in range(5,limit):
		if sieve[p] : P.append(p)
	return P

def containsFactor(itr, n):
	for i in itr:
		if n % i == 0:
			return True
	return False

def isPalindrome(x):
	return str(x) == str(x)[::-1]

def generateTriangleNumbers():
	tri = 0
	i = 1
	while True:
		tri += i
		yield tri
		i += 1
		
def collatzSequence(n):
	seq = [n]
	while seq[-1] != 1:
		if seq[-1] % 2 == 0:
			seq.append(int(seq[-1] / 2))
		else:
			seq.append((seq[-1] * 3) + 1)
	return seq

if __name__ == "__main__":
	print(collatzSequence(13))
		
